tsdb/head_append.go

```diff
@@ -358,6 +384,37 @@ func (a *headAppender) AppendExemplar(ref storage.SeriesRef, lset labels.Labels,
 	return storage.SeriesRef(s.ref), nil
 }
 
+// AppendMetadata for headAppender assumes the series ref already exists, and so it doesn't
+// use getOrCreate or make any of the lset sanity checks that Append does.
+func (a *headAppender) AppendMetadata(ref storage.SeriesRef, lset labels.Labels, meta metadata.Metadata) (storage.SeriesRef, error) {
+	s := a.head.series.getByID(chunks.HeadSeriesRef(ref))
+	if s == nil {
+		s = a.head.series.getByHash(lset.Hash(), lset)
+		if s != nil {
+			ref = storage.SeriesRef(s.ref)
+		}
+	}
+	if s == nil {
+		return 0, fmt.Errorf("unknown series when trying to add metadata with HeadSeriesRef: %d and labels: %s", ref, lset)
+	}
+
+	s.RLock()
+	hasNewMetadata := s.meta != meta
+	s.RUnlock()
+
+	if hasNewMetadata {
+		a.metadata = append(a.metadata, record.RefMetadata{
+			Ref:  s.ref,
+			Type: record.GetMetricType(meta.Type),
```

Comment:
Thanks for bringing this up. This is relevant to your next comment as well.

The enum-like MetricType was added as an encoding/decoding-specific optimisation after [this](https://github.com/prometheus/prometheus/pull/10312#discussion_r810406814) suggestion and it didn't occur to me to make this conversion from the scrape loop.

One drawback I see is that the textparse.MetricType is shared between many structs, so there would a bigger change to review around the scrape loop.  
For example, the current `MetricType` implementation of how metadata arrives to the remote_write would either need an extra conversion, or it would need to propagate the enum-type correctly, resulting in even more changes.